src/Streamly/Internal/Data/Stream/Async/Channel/Dispatcher.hs

```diff
@@ -62,65 +62,102 @@ pushWorker yieldMax sv = do
                     , workerYieldCount = cntRef
                     , workerLatencyStart = lat
                     }
-    doFork (workLoop sv winfo) (svarMrun sv) exception
-        >>= addThread (workerThreads sv)
+    -- In case of lazy dispatch we dispatch workers only from the consumer
+    -- thread. In that case it is ok to use addThread here as it is guaranteed
+    -- that the thread will be added to the workerSet before the thread STOP
+    -- event is processed, because we do both of these actions in the same
+    -- consumer thread. However, in case of eager dispatch we may dispatch
+    -- workers from workers, in which case the thread Stop even may get
+    -- processed before the addThread occurs, so in that case we have to use
+    -- modifyThread which performs a toggle rather than adding or deleting.
+    --
+    -- XXX We can use addThread or modThread based on eager flag.
+    doFork (workLoop sv winfo) (svarMrun sv) exception >>= modThread
 
     where
 
+    modThread = modifyThread (workerThreads sv) (outputDoorBell sv)
     exception = handleChildException (outputQueue sv) (outputDoorBell sv)
 
--- Returns:
--- True: can dispatch more
--- False: cannot dispatch any more
-dispatchWorker :: (MonadIO m, MonadBaseControl IO m) => Count -> Channel m a -> m Bool
-dispatchWorker yieldCount sv = do
+{-# INLINE getEffectiveWorkerLimit #-}
+getEffectiveWorkerLimit :: MonadIO m => Channel m a -> m Limit
```

Comment:
The maximum number of workers required?
Can you explain a little about what this does?